assets/js/src/segments/subscribers_calculator.ts

```diff
@@ -0,0 +1,103 @@
+import MailPoet from 'mailpoet';
+import _ from 'lodash';
+
+export interface FormItem {
+  wordpressRole: string | undefined;
+  segmentType: string | undefined;
+  action: string | undefined;
+  newsletter_id: string | undefined;
+  category_id: string | undefined;
+  product_id: string | undefined;
+}
+
+export interface Result {
+  count: number;
+  errors: string[];
+}
+
+interface ApiVersionWindow extends Window {
+  mailpoet_api_version: string;
+}
+
+declare let window: ApiVersionWindow;
+
+class SubscribersCalculator {
+  emailActions: string[] = ['opened', 'notOpened', 'clicked', 'notClicked'];
+
+  wooCommerceActions: string[] = ['purchasedCategory', 'purchasedProduct'];
+
+  // Names of keys from interface FormItem
+  allowedItemKeys: string[] = ['wordpressRole', 'segmentType', 'action', 'newsletter_id', 'category_id', 'product_id'];
+
+  lastFormItem: FormItem;
+
+  lastResult: Result;
+
+  validateInputData(formItem: FormItem): boolean {
```

Comment:
I am not sure about having this here. The idea, in the beginning, was to keep filter specific logic in the `dynamic_segments_filters` directory. The reason is to keep it in one place and when we are going to add new segments we will have only a couple of files to modify. If we keep adding logic like this, adding new segments will be more complicated. 

The same applies to the actions above.

Reply:
I understand, but how would you solve that we don't want to display the message about calculating and don't call API without all data?

Reply:
I get it. There is a place for some refactoring maybe?

Reply:
I would still have the validation, but move it in the filters. Each filter could validate its own data and return false/true if the data is valid. And having the main validation function that would call dynamically the filter validation based on the type.